### PR TITLE
refactor: Add new concepts for Eigen types

### PR DESCRIPTION
**Pull request by stephenswat**
This commit adds several new concepts for dealing with Eigen base types. The motivation for this is superficially to simplify some of the requirements in the track state proxy and the multi-trajectory, but it also allows me to more easily deal with some clang compiler warnings seen in #3949, as clang won't let me compare `A::ColsAtCompileTime` and `B::ColsAtCompileTime` if `A` and `B` are different types; this allows me to hide the conversion to integers inside a concept.

## Summary by CodeRabbit

- **New Features**
	- Introduced new concepts for Eigen matrix types to enhance type safety and clarity in the `TrackStateProxy` and `VectorMultiTrajectory` classes.
	- Updated method signatures in both classes to utilize these concepts for improved parameter validation.

- **Bug Fixes**
	- Enhanced error handling logic in the `allocateCalibrated_impl` method to ensure consistent exception messaging.

- **Documentation**
	- Added a new header file defining several concepts related to Eigen dense matrix types, improving code expressiveness and type-checking capabilities.

### CODE REVIEW

File: Core/include/Acts/EventData/TrackStateProxy.hpp [hunk 1]
```diff
@@ -782,12 +783,11 @@ class TrackStateProxy {
...
-             Eigen::PlainObjectBase<val_t>::RowsAtCompileTime <= eBoundSize &&
-             Eigen::PlainObjectBase<val_t>::RowsAtCompileTime ==
-                 Eigen::PlainObjectBase<cov_t>::RowsAtCompileTime &&
-             Eigen::PlainObjectBase<cov_t>::RowsAtCompileTime ==
-                 Eigen::PlainObjectBase<cov_t>::ColsAtCompileTime)
+    requires(Concepts::eigen_base_is_fixed_size<val_t> &&
+             Concepts::eigen_bases_have_same_num_rows<val_t, cov_t> &&
+             Concepts::eigen_base_is_square<cov_t> &&
+             Eigen::PlainObjectBase<val_t>::RowsAtCompileTime <=
+                 static_cast<std::underlying_type_t<BoundIndices>>(eBoundSize))
```
Comment (paulgessinger) [inline Core/include/Acts/EventData/TrackStateProxy.hpp:790]: There's a `toUnderlying` helper for this.

File: Core/include/Acts/Utilities/EigenConcepts.hpp [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    static_cast<std::size_t>(Eigen::PlainObjectBase<T1>::RowsAtCompileTime) ==
+        static_cast<std::size_t>(Eigen::PlainObjectBase<T2>::RowsAtCompileTime);
+
+/// @brief Concept that is true iff T1 and T2 have the same, known at compile
+/// time, number of columns.
+template <typename T1, typename T2>
+concept eigen_bases_have_same_num_cols =
+    eigen_base_is_fixed_size<T1> && eigen_base_is_fixed_size<T2> &&
+    static_cast<std::size_t>(Eigen::PlainObjectBase<T1>::ColsAtCompileTime) ==
+        static_cast<std::size_t>(Eigen::PlainObjectBase<T2>::ColsAtCompileTime);
```
Comment (paulgessinger) [inline Core/include/Acts/Utilities/EigenConcepts.hpp:49]: These seem to have fairly limited use outside of the Track Proxy.

Comment (stephenswat) [reply Core/include/Acts/Utilities/EigenConcepts.hpp:49]: Yes this is true, these concepts don't have much stand-alone use by themselves.

On the other hand, we currently have the same `requires` block in three different files which I would like to simplify; as far as I can see we can do it either with concepts like this, or we can come up with a very specific named concept for these `allocateCalibrated` functions, but I'm not sure what we should call that. What do you think?

**Review by paulgessinger — APPROVED** [review]

**Comment by acts-project-service** [comment]
## :white_check_mark: Athena integration test results [26273bcb6dc4979c9f484cc26550cd416932bd0d]
### :white_check_mark: All tests successful

|status|job|report|
|------|---|------|
|:green_circle: | [report_pull_request](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/48503289)||
|:green_circle: | [run_workflow_tests_run4_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/48503288)||
|:green_circle: | [run_workflow_tests_run2_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/48503287)||
|:green_circle: | [run_workflow_tests_run2_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/48503286)||
|:green_circle: | [run_workflow_tests_run3_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/48503285)||
|:green_circle: | [run_workflow_tests_run3_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/48503284)||
|:green_circle: | [test_ActsDumpGeometryIdentifiers](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/48503283)||
|:green_circle: | [test_ActsCheckObjectCountsWorkflowHgtd](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/48503282)||
|:green_circle: | [test_ActsCheckObjectCountsCachedWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/48503281)||
|:green_circle: | [test_ActsCheckObjectCountsWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/48503280)||
|:green_circle: | [test_ActsEFTrackFit](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/48503279)||
|:green_circle: | [test_ActsPersistifySeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/48503278)||
|:green_circle: | [test_ActsBenchmarkWithSpot](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/48503277)||
|:green_circle: | [test_ActsAnalogueClustering](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/48503276)||
|:green_circle: | [test_ActsWorkflowHeavyIons](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/48503275)||
|:green_circle: | [test_ActsWorkflowFastTracking](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/48503274)||
|:green_circle: | [test_ActsWorkflowCached](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/48503273)||
|:green_circle: | [test_ActsWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/48503272)||
|:green_circle: | [test_ActsValidateAmbiguityResolution](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/48503271)||
|:green_circle: | [test_ActsValidateResolvedTracks](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/48503270)||
|:green_circle: | [test_ActsValidateTracks](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/48503269)||
|:green_circle: | [test_ActsValidateActsCoreSpacePoints](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/48503268)||
|:green_circle: | [test_ActsValidateActsSpacePoints](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/48503267)||
|:green_circle: | [test_ActsValidateSeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/48503266)||
|:green_circle: | [test_ActsValidateOrthogonalSeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/48503265)||
|:green_circle: | [test_ActsValidateClusters](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/48503264)||
|:green_circle: | [test_ActsPersistifyEDM](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/48503263)||
|:green_circle: | [test_ActsGx2fRefitting](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/48503262)||
|:green_circle: | [test_ActsGSFRefitting](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/48503261)||
|:green_circle: | [test_ActsKfRefitting](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/48503260)||
|:green_circle: | [test_ActsExtrapolationAlgTest](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/48503259)||
|:green_circle: | [test_ActsITkTest](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/48503258)||

**Pull request merged**